File: TreacheryAndroid/app/src/main/java/com/solomon/treachery/ui/profile/ProfileViewModel.kt
package com.solomon.treachery.ui.profile

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.solomon.treachery.data.FirestoreRepository
import com.solomon.treachery.model.*
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

data class GameStats(
    val totalGames: Int = 0,
    val wins: Int = 0,
    val losses: Int = 0,
    val roleBreakdown: Map<Role, Int> = emptyMap()
) {
    val winRateText: String
        get() = if (totalGames > 0) "${(wins.toDouble() / totalGames * 100).toInt()}%" else "--"
}

@HiltViewModel
class ProfileViewModel @Inject constructor(
    private val firestoreRepository: FirestoreRepository
) : ViewModel() {

    private val _user = MutableStateFlow<TreacheryUser?>(null)
    val user: StateFlow<TreacheryUser?> = _user.asStateFlow()

    private val _gameStats = MutableStateFlow<GameStats?>(null)
    val gameStats: StateFlow<GameStats?> = _gameStats.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    private val _isSaving = MutableStateFlow(false)
    val isSaving: StateFlow<Boolean> = _isSaving.asStateFlow()

    fun loadData(userId: String) {
        viewModelScope.launch {
            _errorMessage.value = null
            try {
                _user.value = firestoreRepository.getUser(userId)
                val games = firestoreRepository.getFinishedGames(userId)
                calculateStats(games, userId)
            } catch (e: Exception) {
                _errorMessage.value = e.localizedMessage
            }
        }
    }

    private suspend fun calculateStats(games: List<Game>, userId: String) {
        var wins = 0
        var losses = 0
        val roleBreakdown = mutableMapOf<Role, Int>()

        for (game in games) {
            try {
                val players = firestoreRepository.getPlayers(game.id)
                val myPlayer = players.find { it.userId == userId }
                val myRole = myPlayer?.role ?: continue

                roleBreakdown[myRole] = (roleBreakdown[myRole] ?: 0) + 1

                val winTeamString = game.winningTeam ?: continue
                val winRole = Role.fromValue(winTeamString) ?: continue
                val didWin = if (winRole == Role.LEADER) {
                    myRole == Role.LEADER || myRole == Role.GUARDIAN
                } else {
                    myRole == winRole
                }
                if (didWin) wins++ else losses++
            } catch (_: Exception) { }
        }

        _gameStats.value = GameStats(
            totalGames = games.size,
            wins = wins,
            losses = losses,
            roleBreakdown = roleBreakdown
        )
    }

    fun saveName(newName: String) {
        val currentUser = _user.value ?: return
        if (newName.isBlank()) return
        _isSaving.value = true
        _errorMessage.value = null

        viewModelScope.launch {
            try {
                val updated = currentUser.copy(displayName = newName.trim())
                firestoreRepository.updateUser(updated)
                _user.value = updated
            } catch (e: Exception) {
                _errorMessage.value = e.localizedMessage
            }
            _isSaving.value = false
        }
    }
}

File: Treachery-iOS/Treachery-iOS/Home/ProfileViewModel.swift
//
//  ProfileViewModel.swift
//  Treachery-iOS
//

import Foundation

struct GameStats {
    let totalGames: Int
    let wins: Int
    let losses: Int
    let roleBreakdown: [Role: Int]

    var winRateText: String {
        guard totalGames > 0 else { return "--" }
        let rate = Double(wins) / Double(totalGames) * 100
        return "\(Int(rate))%"
    }
}

@MainActor
final class ProfileViewModel: ObservableObject {

    // MARK: - Published State

    @Published var user: TreacheryUser?
    @Published var gameStats: GameStats?
    @Published var editedName = ""
    @Published var isEditingName = false
    @Published var isSaving = false
    @Published var errorMessage: String?

    // MARK: - Properties

    private var userId: String?
    private let firestoreManager: FirestoreManaging

    // MARK: - Init

    init(firestoreManager: FirestoreManaging = FirestoreManager()) {
        self.firestoreManager = firestoreManager
    }

    // MARK: - Data Loading

    func loadData(userId: String) async {
        self.userId = userId
        errorMessage = nil

        do {
            async let userTask = firestoreManager.getUser(id: userId)
            async let gamesTask = firestoreManager.getFinishedGames(forUserId: userId)

            user = try await userTask
            let games = try await gamesTask

            await calculateStats(games: games)
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    // MARK: - Name Editing

    func startEditing() {
        editedName = user?.displayName ?? ""
        isEditingName = true
    }

    func saveName() async {
        guard var updatedUser = user, userId != nil else { return }
        guard !editedName.trimmingCharacters(in: .whitespaces).isEmpty else { return }
        isSaving = true
        errorMessage = nil

        updatedUser.displayName = editedName.trimmingCharacters(in: .whitespaces)
        do {
            try await firestoreManager.updateUser(updatedUser)
            user = updatedUser
            isEditingName = false
        } catch {
            errorMessage = error.localizedDescription
        }
        isSaving = false
    }

    // MARK: - Stats Calculation

    private func calculateStats(games: [Game]) async {
        guard let userId else { return }
        var wins = 0
        var losses = 0
        var roleBreakdown: [Role: Int] = [:]

        for game in games {
            if let players = try? await firestoreManager.getPlayers(gameId: game.id),
               let myPlayer = players.first(where: { $0.userId == userId }),
               let myRole = myPlayer.role {

                roleBreakdown[myRole, default: 0] += 1

                if let winTeamString = game.winningTeam,
                   let winRole = Role(rawValue: winTeamString) {
                    let didWin: Bool
                    if winRole == .leader {
                        didWin = myRole == .leader || myRole == .guardian
                    } else {
                        didWin = myRole == winRole
                    }
                    if didWin {
                        wins += 1
                    } else {
                        losses += 1
                    }
                }
            }
        }

        gameStats = GameStats(
            totalGames: games.count,
            wins: wins,
            losses: losses,
            roleBreakdown: roleBreakdown
        )
    }
}
